cmake/pcl_find_sse.cmake

```diff
@@ -1,210 +1,254 @@
-            a = _mm_loadu_pd (vals);
-            b = _mm_hadd_pd (a,a);
-            _mm_storeu_pd (vals, b);
-            return (0);
-        }"
-        HAVE_SSE3_EXTENSIONS)
-
+  if(SSE_LEVEL LESS 2.0)
     if(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX OR CMAKE_COMPILER_IS_CLANG)
-        set(CMAKE_REQUIRED_FLAGS "-msse2")
+      set(CMAKE_REQUIRED_FLAGS "-msse")
     elseif(MSVC AND NOT CMAKE_CL_64)
-        set(CMAKE_REQUIRED_FLAGS "/arch:SSE2")
+      set(CMAKE_REQUIRED_FLAGS "/arch:SSE")
     endif()
 
     check_cxx_source_runs("
-        #include <emmintrin.h>
-        int main ()
-        {
-            __m128d a, b;
-            double vals[2] = {0};
-            a = _mm_loadu_pd (vals);
-            b = _mm_add_pd (a,a);
-            _mm_storeu_pd (vals,b);
-            return (0);
-        }"
-        HAVE_SSE2_EXTENSIONS)
+      #include <xmmintrin.h>
+      int main ()
+      {
+          __m128 a, b;
+          float vals[4] = {0};
+          a = _mm_loadu_ps (vals);
+          b = a;
+          b = _mm_add_ps (a,b);
+          _mm_storeu_ps (vals,b);
+          return (0);
+      }"
+      HAVE_SSE_EXTENSIONS)
 
-    if(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX OR CMAKE_COMPILER_IS_CLANG)
-        set(CMAKE_REQUIRED_FLAGS "-msse")
-    elseif(MSVC AND NOT CMAKE_CL_64)
-        set(CMAKE_REQUIRED_FLAGS "/arch:SSE")
+    if(HAVE_SSE_EXTENSIONS)
+      set(SSE_LEVEL 1.0)
+    endif()
+  endif()
+  
+  if(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX OR CMAKE_COMPILER_IS_CLANG)
+    if(SSE_LEVEL GREATER_EQUAL 4.2)
+      set(SSE_FLAGS "-msse4.2 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 4.1)
+      set(SSE_FLAGS "-msse4.1 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 3.1)
+      set(SSE_FLAGS "-mssse3 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 3.0)
+      set(SSE_FLAGS "-msse3 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 2.0)
+      set(SSE_FLAGS "-msse2 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 1.0)
+      set(SSE_FLAGS "-msse -mfpmath=sse")
+    else()
+      # Setting -ffloat-store to alleviate 32bit vs 64bit discrepancies on non-SSE
+      # platforms.
+      list(APPEND SSE_FLAGS "-ffloat-store")
+    endif()
+  elseif(MSVC AND NOT CMAKE_CL_64)
+    if(SSE_LEVEL GREATER_EQUAL 2.0)
+      set( SSE_FLAGS "/arch:SSE2")
+    elseif(SSE_LEVEL GREATER_EQUAL 1.0)
+      set( SSE_FLAGS "/arch:SSE")
     endif()
+  elseif(MSVC)
+    if(SSE_LEVEL GREATER_EQUAL 4.2)
+      string(APPEND SSE_DEFINITIONS " -D__SSE4_2__")
+    endif()
+    if(SSE_LEVEL GREATER_EQUAL 4.1)
+      string(APPEND SSE_DEFINITIONS " -D__SSE4_1__")
+    endif()
+    if(SSE_LEVEL GREATER_EQUAL 3.1)
+      string(APPEND SSE_DEFINITIONS " -D__SSSE3__")
+    endif()
+    if(SSE_LEVEL GREATER_EQUAL 3.0)
+      string(APPEND SSE_DEFINITIONS " -D__SSE3__")
+    endif()
+    if(SSE_LEVEL GREATER_EQUAL 2.0)
+      string(APPEND SSE_DEFINITIONS " -D__SSE2__")
+    endif()
+    if(SSE_LEVEL GREATER_EQUAL 1.0)
```

Comment:
For Gnu you are using `elseif()`. Isn't the definition of the latest version enough?

Reply:
For Gnu its the compiler flag.
And I by the test I have done, the more secure way is to add all the definitions that are available. Some definitions doesn't seem to be set, if only the compiler flags are used. And therefore some handcraftet guarded code, won't be compiled.

Reply:
I'm actually not fully sure, if this is also the case for linux. I have mainly tested on Windows.

Reply:
You could try  it on godbolt ;-)